sched/scheduler/cluster_state.go

```diff
@@ -12,43 +13,64 @@ const noTask = ""
 type clusterState struct {
 	updateCh chan []cluster.NodeUpdate
 	nodes    map[cluster.NodeId]*nodeState
+	affinity map[string]*nodeSet
+}
+
+type nodeSet struct {
+	idle map[cluster.NodeId]cluster.Node
+	busy map[cluster.NodeId]cluster.Node
+}
+
+func newNodeSet() *nodeSet {
+	return &nodeSet{idle: map[cluster.NodeId]cluster.Node{}, busy: map[cluster.NodeId]cluster.Node{}}
 }
 
 // The State of A Node in the Cluster
 type nodeState struct {
 	node        cluster.Node
 	runningTask string
+	snapshotID  string
 }
 
 // Initializes a Node State for the specified Node
 func newNodeState(node cluster.Node) *nodeState {
 	return &nodeState{
 		node:        node,
 		runningTask: noTask,
+		snapshotID:  "",
 	}
 }
 
 // Creates a New State Distributor with the initial nodes, and which updates
 // nodes added or removed based on the supplied channel.
 func newClusterState(initial []cluster.Node, updateCh chan []cluster.NodeUpdate) *clusterState {
-
 	nodes := make(map[cluster.NodeId]*nodeState)
+	affinity := map[string]*nodeSet{"": newNodeSet()}
 	for _, n := range initial {
 		nodes[n.Id()] = newNodeState(n)
+		affinity[""].idle[n.Id()] = n
 	}
 
-	cs := &clusterState{
+	return &clusterState{
 		updateCh: updateCh,
 		nodes:    nodes,
+		affinity: affinity,
 	}
-
-	return cs
 }
 
 // Update ClusterState to reflect that a task has been scheduled on a
 // particular node
-func (c *clusterState) taskScheduled(nodeId cluster.NodeId, taskId string) {
+func (c *clusterState) taskScheduled(nodeId cluster.NodeId, taskId string, snapshotID string) {
 	ns := c.nodes[nodeId]
+
+	delete(c.affinity[ns.snapshotID].idle, nodeId)
+	delete(c.affinity[ns.snapshotID].busy, nodeId)
```

Comment:
It's not an error but I believe it's redundant wrt the delete in in clusterState.update(). The underlying issue is that I've been seeing nodes re-added to our cluster with a new shardID and then a lagging delete of the old shardID.